### PR TITLE
Add Bangle.JS2 board support

### PR DESCRIPTION
**Pull request by nica-f**
This PR adds board support for the Bangle.JS2 smart watch along with a required driver for its LCD.

Mostly tested by running it on the actual hardware an wearing / using the watch running RIOT 24/7 for many days.

This addresses  #18493

### CODE REVIEW

File: boards/banglejs-2/board.c [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+    gpio_init(ATM_PRESSURE_SDA, GPIO_IN);
+
+    gpio_init(BUTTON0, GPIO_IN_PU);
+    gpio_init(EXTPOWER_PRESENT, GPIO_IN);
+    gpio_init(CHARGE_COMPLETE, GPIO_IN);
+
+    gpio_init(SPI_FLASH_CS, GPIO_OUT);
+    // make sure SPI flash is not selected
+    gpio_set(SPI_FLASH_CS);
+    gpio_init(SPI_FLASH_MISO, GPIO_IN);
```
Comment (benpicco) [inline boards/banglejs-2/board.c:102]: The driver should take care of that

File: boards/banglejs-2/board.c [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+        .driver = &mtd_spi_nor_driver,
+        .page_size = BANGLEJS2_NOR_PAGE_SIZE,
+        .pages_per_sector = BANGLEJS2_NOR_PAGES_PER_SECTOR,
+        .sector_count = BANGLEJS2_NOR_SECTOR_COUNT,
+    },
+    .params = &_banglejs2_nor_params,
+};
+
+mtd_dev_t *mtd0 = (mtd_dev_t *)&banglejs2_nor_dev;
+#endif /* MODULE_MTD */
```
Comment (benpicco) [inline boards/banglejs-2/board.c:64]: If you want to have a file system on the external flash you could add

```C
#ifdef MODULE_VFS_DEFAULT
#include "vfs_default.h"
VFS_AUTO_MOUNT(littlefs2, VFS_MTD(banglejs2_nor_dev), VFS_DEFAULT_NVM(0), 0);
#endif
```

File: boards/banglejs-2/board.c [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+    .wp = GPIO_UNDEF,
+    .hold = GPIO_UNDEF,
+};
+
+static mtd_spi_nor_t banglejs2_nor_dev = {
+    .base = {
+        .driver = &mtd_spi_nor_driver,
+        .page_size = BANGLEJS2_NOR_PAGE_SIZE,
+        .pages_per_sector = BANGLEJS2_NOR_PAGES_PER_SECTOR,
+        .sector_count = BANGLEJS2_NOR_SECTOR_COUNT,
```
Comment (benpicco) [inline boards/banglejs-2/board.c:58]: If you leave this out, the size should be auto-detected on init.

File: boards/banglejs-2/board.c [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+    // LCD CS is high active
+    gpio_init(LCD_CS, GPIO_OUT);
+    gpio_clear(LCD_CS);
+
+    gpio_init(ACCEL_INT, GPIO_IN);
+    gpio_init(ACCEL_SDA, GPIO_IN);
+
+    gpio_init(MAGN_SDA, GPIO_IN);
+    gpio_init(ACCEL_SDA, GPIO_IN);
+    gpio_init(ATM_PRESSURE_SDA, GPIO_IN);
```
Comment (benpicco) [inline boards/banglejs-2/board.c:93]: The driver should take care of this

File: boards/banglejs-2/board.c [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+        pwm_poweroff(PWM_DEV(2));
+    }
+
+    // init ADC
+    adc_init(1); // battery voltage monitor
+}
+
+void board_power_off(void)
+{
+    ztimer_sleep(ZTIMER_MSEC, 5);
```
Comment (benpicco) [inline boards/banglejs-2/board.c:130]: Why is this needed?

File: boards/banglejs-2/Makefile.dep [hunk 6]
```diff
@@ -0,0 +1,18 @@
+# Use Segger's RTT unless another stdio_% is already used
+ifeq (,$(filter stdio_% slipdev_stdio,$(USEMODULE)))
+  USEMODULE += stdio_rtt
+endif
+
+ifneq (,$(filter mtd,$(USEMODULE)))
+  USEMODULE += mtd_spi_nor
+endif
+
+# USEMODULE += nrf5x_periph_shared
```
Comment (benpicco) [inline boards/banglejs-2/Makefile.dep:10]: This does not exist :wink:

File: boards/banglejs-2/Makefile.include [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+# define pyocd as programmer to program with stlink
+PYOCD_FLASH_TARGET_TYPE ?= -t $(CPU)
+
+# this board uses openocd
+PROGRAMMER ?= openocd
+OPENOCD_DEBUG_ADAPTER ?= stlink
+STLINK_VERSION ?= 2
+
+# openocd programmer is supported
+PROGRAMMERS_SUPPORTED += openocd
```
Comment (benpicco) [inline boards/banglejs-2/Makefile.include:13]: Already set by `common/nrf52/Makefile.include`

**Review by benpicco — COMMENTED** [review]
Looks good on a first glance. You probably want to make a separate PR for the display driver.

Two nitpick:

 - coding conventions say you should use `/* C-style comments */` instead of `// C++ style comments`
 - commit messages must be prefixed with the relevent subsystem, so 

    `boards: add Bangle.JS2 smart watch board support`

File: boards/banglejs-2/board.c (hunk 4 above)
Comment (nica-f) [reply boards/banglejs-2/board.c:93]: Right, if the drivers would be in place for these already... so as long as they are not I was thinking to better init the pins to something sane, esp. to not accidentally drive outputs against each other.

Comment (benpicco) [reply boards/banglejs-2/board.c:93]: better add a comment then.

File: boards/banglejs-2/board.c (hunk 5 above)
Comment (nica-f) [reply boards/banglejs-2/board.c:130]: The timer?
Oh dear, I am not sure anymore why I added this, I *think* to make sure that peripherals have shut off before the call to the nRF power off function will just kill everything. Since I do not have the schematics of the watch I have to assume that some of the peripherals like sensors etc. might always be powered to some extent so things need to power themselves off before the nRF goes to sleep. But not really sure of this is really needed, this will for sure be analyzed in more details when I start measuring current consumption and trying to optimize it.

File: boards/banglejs-2/Makefile.dep (hunk 6 above)
Comment (nica-f) [reply boards/banglejs-2/Makefile.dep:10]: Right, not anymore, will remove it entirely, just slipped my attention.

File: boards/banglejs-2/board.c (hunk 5 above)
Comment (benpicco) [reply boards/banglejs-2/board.c:130]: But nothing happens *before* the sleep - so what effect does it have?

**Comment by nica-f** [comment]
OK, I am doing this again and then splitting the PR into board support and LCD driver... closing.
